gpflow/optimizers/scipy.py

```diff
@@ -34,27 +36,44 @@ def minimize(self,
         if not callable(closure):
             raise TypeError('Callable object expected.')  # pragma: no cover
         initial_params = self.initial_parameters(variables)
-        func = self.eval_func(closure, variables, step_callback)
-        return scipy.optimize.minimize(func, initial_params, jac=True, **scipy_kwargs)
+        func = self.eval_func(closure, variables)
+
+        if step_callback is not None:
+            if 'callback' in scipy_kwargs:
+                raise ValueError("Callback passed both via `step_callback` and `callback`")
+
+            callback = self.callback_func(closure, variables, step_callback)
+            scipy_kwargs.update(dict(callback=callback))
+
+        method = scipy_kwargs.pop('method', self._DEFAULT_METHOD)
+        return scipy.optimize.minimize(func, initial_params, jac=True, method=method,
+                                       **scipy_kwargs)
 
     @classmethod
     def initial_parameters(cls, variables):
         return cls.pack_tensors(variables)
 
     @classmethod
-    def eval_func(cls, closure: LossClosure, variables: Variables, step_callback: Optional[StepCallback] = None):
+    def eval_func(cls, closure: LossClosure, variables: Variables):
+        def _eval(x):
+            cls.unpack_tensors(variables, x)
+            loss, grads = _compute_loss_and_gradients(closure, variables)
+            return loss.numpy().astype(np.float64), cls.pack_tensors(grads).astype(np.float64)
+
+        return _eval
+
+    @classmethod
+    def callback_func(cls, closure: LossClosure, variables: Variables, step_callback: Optional[StepCallback] = None):
         step = 0  # type: int
 
-        def _eval(x):
+        def _callback(x):
             nonlocal step
             cls.unpack_tensors(variables, x)
             loss, grads = _compute_loss_and_gradients(closure, variables)
-            if step_callback is not None:
-                step_callback(step=step, loss=loss, variables=variables, gradients=grads)
+            step_callback(step=step, loss=loss, variables=variables, gradients=grads)
```

Comment:
so by splitting out the callback from the eval function, we need to compute the loss and gradients twice. is there anyway to avoid this? i suspect not, but this may well seriously slow down, depending on how computationally heavy it is to calc the objective (is there memoization?)

Reply:
if not does it make any sense to attempt some memoization?

Reply:
**Proposal**:
1. We keep the same function as before: that is the `eval_func` has the same signature `def eval_func(cls, closure, variables, step_callback)`
2. In the `eval_func` we have non_local variables which include the `loss` and `gradient`, and we proceed to make not one, but _two_ closures - just as you have done in your change. One for eval, one for callback.
3. The `eval` closure now also writes to the `non_local` `loss` and `gradient`. The `callback` closure now only reads from these values. 
4. Since the variables and `non_local`, they can't leak outside the closure (without serious work), and avoid global state, while saving on the function call.

Reply:
Neat idea but I don't think would work as the last call to eval_func during a line search may not be the one that gives the optimal value.

Changed the step_callback interface so that instead of recomputing loss&gradient every step(!), to only pass variables and (unpacked) values (as tensors) so that user can assign as required (or not).